### PR TITLE
Fix #11752: fix numerical issue in log_softmax

### LINKED ISSUE #11752: nn.CrossEntropyLoss() yields wrong output for big logits
## Issue description

`nn.CrossEntropyLoss()` yields incorrect results for big logits.

## Code example

``` python
import torch
from torch import nn

crit = nn.CrossEntropyLoss()

target = torch.ones(1, dtype=torch.long)

y_small = torch.ones(1, 2, dtype=torch.float32)
y_big = y_small * 1e8

loss_small = crit(y_small, target)
loss_big = crit(y_big, target)
loss_fixed = crit(y_big - y_big.max(), target)

print('loss_small = {}\n'
      'loss_big   = {}\n'
      'loss_fixed = {}'.format(
        loss_small, loss_big, loss_fixed))
```

`loss_small` should be equal to `loss_big`, as both `y_small` and `y_big` contain logits that correspond to the same distribution. However, the output of the above is:

```
loss_small = 0.6931471824645996
loss_big   = 0.0
loss_fixed = 0.6931471824645996
```

Note that the output is correct (`loss_fixed == loss_small`) if I apply the trick `softmax(x) = softmax(x - c)` for some `c`, in particular, for `c=max(x)`.

This happens on CPU and GPU.

## System Info

PyTorch version: 0.4.1
Is debug build: No
CUDA used to build PyTorch: 9.0.176

OS: Debian GNU/Linux 8 (jessie)
GCC version: (Debian 4.9.2-10+deb8u1) 4.9.2
CMake version: version 3.0.2

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 9.0.176
GPU models and configuration: 
GPU 0: TITAN Xp

Nvidia driver version: 384.130
cuDNN version: Could not collect

Versions of relevant libraries:
[pip] numpy (1.14.3)
[pip] torch (0.4.1)
[pip] torchvision (0.2.1)
[conda] pytorch                   0.4.1           py36_cuda9.0.176_cudnn7.1.2_1    pytorch
[conda] torchvision               0.2.1                    py36_1    pytorch

### PR DESCRIPTION
**Pull request by wolegechu**
#11866 has corrected this issue in function `host_softmax` (aten/src/ATen/native/SoftMax.cpp). But I tried the example proposed in #11752. `log_softmax` is still not working for big logits.

I have looked into the source code, found that example had called `vec_host_softmax_lastdim`, not `host_softmax`.

This code fixes the issue in `_vec_log_softmax_lastdim` and has a test for `log_softmax`.

### CODE REVIEW

File: test/test_nn.py [hunk 1]
```diff
@@ -7847,6 +7848,11 @@ def test_softmin(self):
         self.assertEqual(F.softmin(x, 1), F.softmax(-x, 1))
         self.assertEqual(F.softmin(x, 0), F.softmax(-x, 0))
 
+    def test_log_softmax(self):
+        x_small = torch.ones(1, 2, dtype=torch.float32)
+        x_big = x_small * 1e16
```
Comment (vadimkantorov) [inline test/test_nn.py]: softmax and logsoftmax should be invariant uner addition, but not multiplication, right? (multilicative factor is inverse temperature, right?)

should this line read  `x_big = x_small + 1e16`?

Comment (wolegechu) [reply test/test_nn.py]: oops..., this test works in the special case, but it's a mistake. I've written a new commit.

File: aten/src/ATen/native/cpu/SoftMaxKernel.cpp [hunk 2]
```diff
@@ -65,19 +65,19 @@ inline void _vec_log_softmax_lastdim(
...
               loop_end);
           for (int64_t j = 0; j < loop_end; j++) {
             int64_t i = ii + j;
             scalar_t* input_data = input_data_base + i * dim_size;
             scalar_t* output_data = output_data_base + i * dim_size;
             scalar_t tmp_sum = tmp_sum_scalar[j];
+            scalar_t max_input = max_input_arr[j];
             vec256::map(
-                [tmp_sum](Vec x) { return x - Vec(tmp_sum); },
+                [tmp_sum, max_input](Vec x) { return x - Vec(max_input) - Vec(tmp_sum); },
```
Comment (SsnL) [inline aten/src/ATen/native/cpu/SoftMaxKernel.cpp]: Does the compiler optimize this into computing `max_input` and `tmp_sum` before the `:map`?

Comment (wolegechu) [reply aten/src/ATen/native/cpu/SoftMaxKernel.cpp]: It shouldn’t be optimized to compute before. 

In some cases that input are large digits and their differences is small, the preprocessed input value would be ignored. You can see the example here.https://github.com/pytorch/pytorch/issues/11752#issuecomment-422883379

And I have look up the `log_sofmax` implementation MXNet and TensorFlow. They also write `x - Vec(max_input) - Vec(tmp_sum)` together to ensure the computing order.

Comment (SsnL) [reply aten/src/ATen/native/cpu/SoftMaxKernel.cpp]: Could you add a comment here so people won't "optimize" this away in future? Thanks!

File: test/test_nn.py [hunk 3]
```diff
@@ -821,7 +821,7 @@ def test_invalid_conv1d(self):
             module(input)
 
     def test_invalid_conv2d(self):
-        module = torch.nn.Conv2d(1, 1, kernel_size=3, dilation=2, stride=2)
+        module = nn.Conv2d(1, 1, kernel_size=3, dilation=2, stride=2)
```
Comment (VitalyFedyunin) [inline test/test_nn.py]: Could you please remove all unrelated changes.
It is much harder to track code history if we touch unrelated things.

**Review by wolegechu — COMMENTED** [review]
@VitalyFedyunin So strange... I merged the 'test/test_nn.py' file from upstream:master branch to my branch. Now Github highlights the codes in green that don't belong to me.

File: test/test_nn.py (hunk 3 above)
Comment (wolegechu) [reply test/test_nn.py]: It's done.

Comment (wolegechu) [reply test/test_nn.py]: I tried and I'm in a mess...

I think it's not very necessary...

File: aten/src/ATen/native/cpu/SoftMaxKernel.cpp (hunk 2 above)
Comment (wolegechu) [reply aten/src/ATen/native/cpu/SoftMaxKernel.cpp]: It's done.

**Comment by wolegechu** [comment]
> @VitalyFedyunin So strange... I merged the 'test/test_nn.py' file from upstream:master branch to my branch. Now Github highlights the codes in green that don't belong to me.
I introduced some new commits, which had been merged into master after I created this PR.

@VitalyFedyunin emmmmm...Anyway, I  have removed all unrelated changes.

File: test/test_nn.py (hunk 3 above)
Comment (wolegechu) [reply test/test_nn.py]: ....It's done.

**Review by VitalyFedyunin — APPROVED** [review]

**Review by VitalyFedyunin — APPROVED** [review]
